src/main/java/build/buildfarm/instance/memory/MemoryInstance.java

```diff
@@ -153,15 +153,26 @@ protected void enqueueOperation(Operation operation) {
 
   @Override
   protected void updateOperationWatchers(Operation operation) {
-    Iterable<Consumer<Operation>> operationWatchers =
+    List<Function<Operation, Boolean>> operationWatchers =
         watchers.get(operation.getName());
     if (operationWatchers != null) {
       if (operation.getDone()) {
         watchers.remove(operation.getName());
       }
+      long unfilteredWatcherCount = operationWatchers.size();
       super.updateOperationWatchers(operation);
-      for (Consumer<Operation> watcher : operationWatchers) {
-        watcher.accept(operation);
+      ImmutableList.Builder<Function<Operation, Boolean>> filteredWatchers = new ImmutableList.Builder<>();
+      long filteredWatcherCount = 0;
+      for( Function<Operation, Boolean> watcher : operationWatchers ) {
```

Comment:
Added ca1b5d839433cccc7d8b9a866eb5208fe572d4a1 to deal with this one, and fixed it for the subsequent additions.